pkg/controller/registry/resolver/installabletypes.go

```diff
@@ -55,32 +56,110 @@ func bundleId(bundle, channel string, catalog cache.SourceKey) solver.Identifier
+type ConstraintProviderFunc func(o *cache.Operator) ([]solver.Constraint, error)
+
+func (c ConstraintProviderFunc) Constraints(o *cache.Operator) ([]solver.Constraint, error) {
+	return c(o)
+}
+
+// constraintProviderList provides aggregate constraints from a list of ConstraintProviders.
+type constraintProviderList struct {
+	mu        sync.RWMutex
+	providers []ConstraintProvider
+}
+
+// add appends the given ConstraintProviders to the list aggregated over by the constraintProviderList.
+// add is threadsafe.
+func (c *constraintProviderList) add(providers ...ConstraintProvider) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.providers = append(c.providers, providers...)
+}
+
+func (c *constraintProviderList) Constraints(o *cache.Operator) ([]solver.Constraint, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
 	var constraints []solver.Constraint
-	if o.SourceInfo.Catalog.Virtual() && o.SourceInfo.Subscription == nil {
+	for _, provider := range c.providers {
+		cons, err := provider.Constraints(o)
+		if err != nil {
+			return nil, err
+		}
+
+		constraints = append(constraints, cons...)
+	}
+
+	return constraints, nil
+}
+
+var (
+	// systemConstraintProviders is the list of constraint providers used by all solvers.
+	systemConstraintProviders = constraintProviderList{
+		providers: []ConstraintProvider{
+			freestandingCSVConstraint(),
+			deprecatedConstraint(),
+		},
+	}
+)
+
+func freestandingCSVConstraint() ConstraintProviderFunc {
+	return func(o *cache.Operator) ([]solver.Constraint, error) {
+		if !(o.SourceInfo.Catalog.Virtual() && o.SourceInfo.Subscription == nil) {
+			return nil, nil
+		}
+
 		// CSVs already associated with a Subscription
 		// may be replaced, but freestanding CSVs must
 		// appear in any solution.
-		constraints = append(constraints, PrettyConstraint(
+		return []solver.Constraint{PrettyConstraint(
 			solver.Mandatory(),
 			fmt.Sprintf("clusterserviceversion %s exists and is not referenced by a subscription", o.Name),
-		))
-	}
-	for _, p := range o.Properties {
-		if p.GetType() == operatorregistry.DeprecatedType {
-			constraints = append(constraints, PrettyConstraint(
-				solver.Prohibited(),
-				fmt.Sprintf("bundle %s is deprecated", id),
-			))
-			break
+		)}, nil
+	}
+}
+
+func deprecatedConstraint() ConstraintProviderFunc {
+	return func(o *cache.Operator) ([]solver.Constraint, error) {
+		id := bundleId(o.Name, o.Channel(), o.SourceInfo.Catalog)
+		for _, p := range o.Properties {
+			if p.GetType() == operatorregistry.DeprecatedType {
+				return []solver.Constraint{PrettyConstraint(
+					solver.Prohibited(),
+					fmt.Sprintf("bundle %s is deprecated", id),
+				)}, nil
+			}
 		}
+
+		return nil, nil
 	}
+}
+
+// AddSystemConstraintProviders adds providers to the list of providers used system-wide, across all solvers.
+func AddSystemConstraintProviders(providers ...ConstraintProvider) {
+	systemConstraintProviders.add(providers...)
+}
```

Comment:
Is this the extension point were callers can add custom providers (e.g. from an `init` function)?

It might be good to add an example in the PR summary of how a caller would add a custom provider.